File: common/src/main/java/org/bread_experts_group/breadlib/platform/PlatformServices.java
package org.bread_experts_group.breadlib.platform;

import org.bread_experts_group.breadlib.BreadLib;

import java.util.ServiceLoader;

public class PlatformServices {
	public static final IPlatformHelper PLATFORM = load(IPlatformHelper.class);

	public static <T> T load(Class<T> clazz) {
		final T loadedService = ServiceLoader.load(clazz).findFirst().orElseThrow(() -> new NullPointerException("Failed to load service for " + clazz.getName()));
		BreadLib.LOGGER.debug("Loaded {} for service {}", loadedService, clazz);
		return loadedService;
	}
}

File: common/src/main/kotlin/org/bread_experts_group/breadlib/platform/PlatformServices.kt
package org.bread_experts_group.breadlib.platform

import org.bread_experts_group.breadlib.BreadLib
import java.util.*

object PlatformServices {
	@JvmField
	val PLATFORM: IPlatformHelper = load(IPlatformHelper::class.java)
	val NETWORK: NetworkHelper = load(NetworkHelper::class.java)
	val CAPABILITY: ICapabilityHelper = load(ICapabilityHelper::class.java)

	private fun <T> load(clazz: Class<T>): T {
		val loadedService = ServiceLoader.load(clazz).findFirst()
			.orElseThrow { NullPointerException("Failed to load service for " + clazz.getName()) }
		BreadLib.LOGGER.debug("Loaded {} for service {}", loadedService, clazz)
		return loadedService
	}
}